core/good_key_test.go

```diff
@@ -84,3 +79,112 @@ func TestGoodKey(t *testing.T) {
+		test.AssertError(t, GoodKey(private.PublicKey), "Should not have accepted key not on the curve.")
+	}
+}
+
+func TestECDSANotOnCurveY(t *testing.T) {
+	for _, curve := range validCurves {
+		// Again with Y.
+		private, err := ecdsa.GenerateKey(curve, rand.Reader)
+		test.AssertNotError(t, err, "Error generating key")
+
+		// Change the public key so that it is no longer on the curve.
+		private.Y.Add(private.Y, big.NewInt(1))
+		test.AssertError(t, GoodKey(&private.PublicKey), "Should not have accepted key not on the curve.")
+		test.AssertError(t, GoodKey(private.PublicKey), "Should not have accepted key not on the curve.")
+	}
+}
+
+func TestECDSANegative(t *testing.T) {
+	for _, curve := range validCurves {
+		// Check that negative X is not accepted.
+		private, err := ecdsa.GenerateKey(curve, rand.Reader)
+		test.AssertNotError(t, err, "Error generating key")
+
+		private.X.Neg(private.X)
+		test.AssertError(t, GoodKey(&private.PublicKey), "Should not have accepted key with negative X.")
+		test.AssertError(t, GoodKey(private.PublicKey), "Should not have accepted key with negative X.")
+
+		// Check that negative Y is not accepted.
+		private.X.Neg(private.X)
+		private.Y.Neg(private.Y)
+		test.AssertError(t, GoodKey(&private.PublicKey), "Should not have accepted key with negative Y.")
+		test.AssertError(t, GoodKey(private.PublicKey), "Should not have accepted key with negative Y.")
+	}
+}
+
+func TestECDSANegativeUnmodulatedX(t *testing.T) {
+	for _, curve := range validCurves {
+		// Check that unmodulated X is not accepted.
+		private, err := ecdsa.GenerateKey(curve, rand.Reader)
+		test.AssertNotError(t, err, "Error generating key")
+
+		private.X.Mul(private.X, private.Curve.Params().P)
+		test.AssertError(t, GoodKey(&private.PublicKey), "Should not have accepted key with unmodulated X.")
+		test.AssertError(t, GoodKey(private.PublicKey), "Should not have accepted key with unmodulated X.")
+	}
+}
+
+func TestECDSANegativeUnmodulatedY(t *testing.T) {
+	for _, curve := range validCurves {
+		// Check that unmodulated Y is not accepted.
+		private, err := ecdsa.GenerateKey(curve, rand.Reader)
+		test.AssertNotError(t, err, "Error generating key")
+
+		private.X.Mul(private.Y, private.Curve.Params().P)
+		test.AssertError(t, GoodKey(&private.PublicKey), "Should not have accepted key with unmodulated Y.")
+		test.AssertError(t, GoodKey(private.PublicKey), "Should not have accepted key with unmodulated Y.")
+	}
+}
+
+func TestECDSAIdentity(t *testing.T) {
+	for _, curve := range validCurves {
+		// The point at infinity is 0,0, it should not be accepted.
+		public := ecdsa.PublicKey{
+			Curve: curve,
+			X:     big.NewInt(0),
+			Y:     big.NewInt(0),
+		}
+
+		test.AssertError(t, GoodKey(&public), "Should not have accepted key with point at infinity.")
+		test.AssertError(t, GoodKey(public), "Should not have accepted key with point at infinity.")
+	}
+}
```

Comment:
1) TestECDSANotOnCurveY so there is an check for an point not on the curve.
2) The check for additiveIdentity is ok, because 2 * (0,0) = (0,0) and if there would
   be any other point that is going to itself it is also not useable as an private key.

> > 2*P = 0 is not sufficient evidence that the point is the identity

> func isAdditiveIdentity(c elliptic.Curve, x, y *big.Int) bool {
> 	// The point at infinity (and additive identity) for all supported curves is
> 	// (0,0). But we can implement this more generally by satisfying ourselves
> 	// that the values provided are not the additive identity.
> 	ax, ay := c.Add(x, y, x, y)
> 	return ax.Cmp(x) == 0 && ay.Cmp(y) == 0
> }